### ISSUE TITLE
More flexible loose application configuration?

### ISSUE THREAD

**Issue by hrstoyanov** [WASdev/ci.gradle#167]
The [Loose Application](https://www.ibm.com/support/knowledgecenter/en/SS7K4U_liberty/com.ibm.websphere.wlp.zseries.doc/ae/rwlp_loose_applications.html) idea is a very clever one and allows to cut down the redeploy time during development a lot. Unfortunately, this Gradle plug-in makes some assumptions about how to construct the loose app XML file and provides no flexibility. 

To illustrate the issue, here are the "tricks" I have to do when constructing my WAR in order to deal with some bad dependencies:

```
//Only add jar files to the war that we know we will need at runtime!
war {
    duplicatesStrategy 'exclude'
    classpath configurations.warLib
    classpath = classpath.filter { file ->
        (!file.name.startsWith('javaee-api')
                && !file.name.startsWith('javax')
                && !file.name.startsWith('junit')
                && !file.name.startsWith('hamcrest')
                && !file.name.startsWith('gwt-slf4j')
                && !file.name.startsWith('dom4j')
                && !file.name.startsWith('guava-gwt')
                && !file.name.startsWith('jgroups')
                && !file.name.startsWith('j2objc')
                && !file.name.startsWith('error_prone')

                //Exclude Errai client side jars
                && !file.name.startsWith('errai-cdi-client')
                //&& !file.name.startsWith('errai-jaxrs-client')
                && !file.name.startsWith('errai-navigation')
                //&& !file.name.startsWith('errai-codegen')
                && !file.name.startsWith('errai-ui')
                && !file.name.startsWith('errai-data-binding')
        )
    }
    rootSpec.exclude('**/com/recres/web/gwt/client/')
}
```

In order for the loose app approach to work for me, this plug-in needs to provide similar level of control of what makes up the app. 

Are thee any plans to enhance this plug-in or has anyone deallth with WARs that are not exactly trivial?

Thanks

**Comment by jjvilleg** [comment]
Have you tried adding them as provided dependencies? i.e. providedCompile or providedRuntime? This will keep the dependencies in the same scope, but keep them from being packaged in the WAR.

https://docs.gradle.org/current/userguide/war_plugin.html#sec:war_dependency_management

**Comment by hrstoyanov** [comment]
Yes, I have tried ```providedCompile``` and ```providedRuntime```. The issue is that if artifact A.jar is in both ```compile``` and ```providedCompile```, very strange things happen to its transitive dependencies (A.jar->B.jar), which in my case is not what I wanted.

**Comment by jjvilleg** [comment]
What is the reason for having artifact A.jar as both a `compile` and a `providedCompile` dependency?

How are you configuring the war plugin to pick up the js code that is not in the `src/main/webapp` directory? Can you provide an example of how you are configuring your war plugin?

**Comment by hrstoyanov** [comment]
I do need to deal with some bad dependencies in both ```compile``` and ``providedCompile```. It is surprising how Gradle treats them though, I think ```providedCompile``` takes precedence and removes all other transitives.


Here is how one can include additional directories with the WAR plugin:

```
war {
    from 'src/rootContent' // adds a file-set to the root of the archive
    webInf { from 'src/additionalWebInf' } // adds a file-set to the WEB-INF dir.
    classpath fileTree('additionalLibs') // adds a file-set to the WEB-INF/lib dir.
    classpath configurations.moreLibs // adds a configuration to the WEB-INF/lib dir.
    webXml = file('src/someWeb.xml') // copies a file to WEB-INF/web.xml
}
```

**Comment by hrstoyanov** [comment]
Here is a better example of the gap in this plugin:
```
war {
    from ("${projectDir}") {
        include ('src/main/java/**')
    }
}
```

The above WAR configuration adds an additional folder (could be anything, for example generated JS code) to the root of the war file - open the produced war bundle and you will see the inclusion. However, the generated loose application XML ignores the additional entry and you only get the default this:
```
<?xml version="1.0" encoding="UTF-8"?>
<archive>
    <dir sourceOnDisk="...../my-webapp/src/main/webapp" targetInArchive="/"/>
...
</archive>
```
```
war {
    from ("${projectDir}") {
        include ('src/main/java/**')
    }
}
```

Another example:
```
war {
    rootSpec.exclude("**/my.jar")
    rootSpec.exclude("**/my/special/*.class")
}
```

The above configuration will exclude ```my.war``` and part of the classes from the war bundle. None of this is happening with the loose application xml generator!

I tried to alert the openliberty team that the XML format is missing exclusion, but they did not this this is important (see [this](https://github.com/OpenLiberty/open-liberty/issues/11080) ).

**Comment by hrstoyanov** [comment]
?

**Comment by cherylking** [comment]
We feel with the complexity behind this (needing to keep track of the extra files/folders and any changes/adds/deletes to them), this is not something we want to implement. This is a gradle limitation.

**Comment by Emily-Jiang** [comment]
@cherylking can you please open an issue to report the gradle limitation and reopen this issue to track that change?

**Comment by cherylking** [comment]
Reopening so we can revisit this behavior.
